django/db/models/constraints.py

```diff
@@ -373,21 +395,28 @@ def validate(self, model, instance, exclude=None, using=DEFAULT_DB_ALIAS):
         if not self.condition:
             if queryset.exists():
                 if self.expressions:
-                    raise ValidationError(self.get_violation_error_message())
+                    raise ValidationError(
+                        self.get_violation_error_message(),
+                        code=self.violation_error_code,
+                    )
                 # When fields are defined, use the unique_error_message() for
                 # backward compatibility.
                 for model, constraints in instance.get_constraints():
                     for constraint in constraints:
                         if constraint is self:
                             raise ValidationError(
-                                instance.unique_error_message(model, self.fields)
+                                instance.unique_error_message(model, self.fields),
+                                code=self.violation_error_code,
```

Comment:
Inside this PR I think you're right, it would make more sense to be consistent with `violation_error_message`.

But in the long run, I think it would make more sense to use the user provided `violation_error_code` & `violation_error_message` if they exist (and fallback on `unique_error_message` behavior otherwise).

Because currently (if I'm not mistaken), if you define a `UniqueConstraint(fields=('some_field',))` or `UniqueConstraint(fields=('some_field',), violation_error_message="never used")` you end up with the same behavior where `violation_error_message="never used"` is never used ?

Reply:
> But in the long run, I think it would make more sense to use the user provided `violation_error_code` & `violation_error_message` if they exist (and fallback on `unique_error_message` behavior otherwise).
> 
> Because currently (if I'm not mistaken), if you define a `UniqueConstraint(fields=('some_field',))` or `UniqueConstraint(fields=('some_field',), violation_error_message="never used")` you end up with the same behavior where `violation_error_message="never used"` is never used ?
> 
> If the user bothered to provide `violation_error_code` & `violation_error_message`, they should be used.

It's a [documented](https://docs.djangoproject.com/en/4.1/ref/models/constraints/#id1) behavior that we introduced for backward compatibility. We're also working on improving classification in other cases, see #16054. We'll probably change the default value of `UniqueConstraint.violation_error_code` to `unique` in #16054, after merging this PR. 

\cc @shangxiao

Reply:
I removed this part.

I'm considering making a followup PR to better handle this special-case (basically calling `unique_error_message` in `UniqueConstraint.__init__` if `self.condition` and `self.expressions` are empty).

Reply:
I've submitted #16598 as followup to allow customizing the code, even for `UniqueConstraint` with fields and no condition.